naga/tests/validation.rs

```diff
@@ -311,3 +311,180 @@ fn main() {{
+fn no_flat_first_in_glsl() {
+    let DummyInterpolationShader {
+        source: _,
+        module,
+        interpolate_attr,
+        entry_point,
+    } = DummyInterpolationShader::new(naga::Interpolation::Flat, Some(naga::Sampling::First));
+
+    let mut validator = naga::valid::Validator::new(Default::default(), Default::default());
+    let module_info = validator.validate(&module).unwrap();
+
+    let options = Default::default();
+    let pipeline_options = naga::back::glsl::PipelineOptions {
+        shader_stage: naga::ShaderStage::Fragment,
+        entry_point: entry_point.to_owned(),
+        multiview: None,
+    };
+    let mut glsl_writer = naga::back::glsl::Writer::new(
+        String::new(),
+        &module,
+        &module_info,
+        &options,
+        &pipeline_options,
+        Default::default(),
+    )
+    .unwrap();
+
+    let err = glsl_writer.write().expect_err(&format!(
+        "`{interpolate_attr}` should fail backend validation"
+    ));
+
+    assert!(matches!(
+        err,
+        naga::back::glsl::Error::FirstSamplingNotSupported
+    ));
+}
+
+struct DummyInterpolationShader {
+    source: String,
+    module: naga::Module,
+    interpolate_attr: String,
+    entry_point: &'static str,
+}
+
+impl DummyInterpolationShader {
+    fn new(interpolation: naga::Interpolation, sampling: Option<naga::Sampling>) -> Self {
+        // NOTE: If you have to add variants below, make sure to add them to the
+        // `cartesian_product`'d combinations in tests around here!
+        let interpolation_str = match interpolation {
+            naga::Interpolation::Flat => "flat",
+            naga::Interpolation::Linear => "linear",
+            naga::Interpolation::Perspective => "perspective",
+        };
+        let sampling_str = match sampling {
+            None => String::new(),
+            Some(sampling) => format!(
+                ", {}",
+                match sampling {
+                    naga::Sampling::First => "first",
+                    naga::Sampling::Either => "either",
+                    naga::Sampling::Center => "center",
+                    naga::Sampling::Centroid => "centroid",
+                    naga::Sampling::Sample => "sample",
+                }
+            ),
+        };
+        let member_type = match interpolation {
+            naga::Interpolation::Perspective | naga::Interpolation::Linear => "f32",
+            naga::Interpolation::Flat => "u32",
+        };
+
+        let interpolate_attr = format!("@interpolate({interpolation_str}{sampling_str})");
+        let source = format!(
+            "\
+struct VertexOutput {{
+    @location(0) {interpolate_attr} member: {member_type},
+}}
+
+@fragment
+fn main(input: VertexOutput) {{
+    // ...
+}}
+"
+        );
+        let module = naga::front::wgsl::parse_str(&source).unwrap();
+
+        Self {
+            source,
+            module,
+            interpolate_attr,
+            entry_point: "main",
+        }
+    }
+}
```

Comment:
@teoxoy: I don't think this is too interesting, but just wanted to give you a heads-up that I extracted this `DummyInterpolationShader` API from the now-two tests being introduced with 0591c9a5.